### PR TITLE
Feature: Add Options for Text Diff

### PR DESCRIPTION
**Pull request by NirvanaPhX**
Hi Dan, 

I have included most of the diff options but the diffArray method was having some troubles with TypeScript.

Also I am having some troubles with testing the <Select> MUI element.
It cannot find that element.
I have tried adding "data-testid" inside the inputProps of the element and tried to obtain the element with getByTestId.
I have also tried to use the getByLabelText which doesn't work either.

Please let me know if I am doing anything wrong here.

Thanks

### CODE REVIEW

File: __TESTS__/text-diff.spec.tsx [hunk 1]
```diff
@@ -6,13 +6,12 @@ import React from 'react';
...
 import TextDiff from '../pages/text-diff';
 
-const testInput1 = 'I am the Lorax';
-const testInput2 = 'I speak for the trees';
-const testOutput =
-  '<span style="color:grey">I </span><span style="color:green">spe</span><span style="color:grey">a</span><span style="color:red">m</span><span style="color:green">k</span><span style="color:grey"> </span><span style="color:green">for </span><span style="color:grey">the </span><span style="color:red">Lo</span><span style="color:green">t</span><span style="color:grey">r</span><span style="color:red">ax</span><span style="color:green">ees</span>';
-
 describe('TextDiff', () => {
   it('Shows different characters', async () => {
+    const testInput1 = 'I am the Lorax';
```
Comment (dlford) [inline __TESTS__/text-diff.spec.tsx:11]: Please put these back where they were, the indentation breaks the existing tests, we shouldn't need to change those

File: pages/text-diff.tsx [hunk 2]
```diff
@@ -66,13 +144,48 @@ export default function TextDiffPage() {
 
   return (
     <Layout title='Text Difference'>
-      <Heading>Text Diff</Heading>
+      <Heading>Text Diff {selectedOptions?.label}</Heading>
```
Comment (dlford) [inline pages/text-diff.tsx:147]: This can cause layout shift, lets take it out please

File: pages/text-diff.tsx [hunk 3]
```diff
@@ -66,13 +144,48 @@ export default function TextDiffPage() {
...
+        flexDirection='column'
+        justifyContent='stretch'
+        width={1000}
+        maxWidth='100%'
+        gap={2}
+        mb={3}
+      >
+        <FormControl>
+          <InputLabel id='diff-select-label'>
+            Choose Diff Options
```
Comment (dlford) [inline pages/text-diff.tsx:165]: This is what you should use in `getByLabelText` ("Choose Diff Options")

As a last resort you can add `data-testid="some-unique-string"` to any element, but it is better to use the other queries (e.g. `getByLabelText`) when possible

File: __TESTS__/text-diff.spec.tsx [hunk 4]
```diff
@@ -27,4 +26,40 @@ describe('TextDiff', () => {
...
+    const testInput2 = 'I speak for the trees';
+    const testOutput =
+      '<span style="color:grey">I </span><span style="color:green">spe</span><span style="color:grey">a</span><span style="color:red">m</span><span style="color:green">k</span><span style="color:grey"> </span><span style="color:green">for </span><span style="color:grey">the </span><span style="color:red">Lo</span><span style="color:green">t</span><span style="color:grey">r</span><span style="color:red">ax</span><span style="color:green">ees</span>';
+    const user = userEvent.setup();
+    render(<TextDiff />);
+
+    const input1 = screen.getByLabelText(/Text 1/i);
+    const input2 = screen.getByLabelText(/Text 2/i);
+    const output = screen.getByTestId('text-difference-output');
```
Comment (dlford) [inline __TESTS__/text-diff.spec.tsx:56]: Try this: `getByLabelText(/Choose Diff Options/i);`

File: pages/text-diff.tsx [hunk 5]
```diff
@@ -4,16 +4,87 @@ import Box from '@mui/material/Box';
 import Button from '@mui/material/Button';
 import TextField from '@mui/material/TextField';
 import Typography from '@mui/material/Typography';
+import MenuItem from '@mui/material/MenuItem';
```
Comment (dlford) [inline pages/text-diff.tsx:7]: linter errors here, run `npm run lint` to fix/check

File: pages/text-diff.tsx [hunk 6]
```diff
@@ -33,12 +104,19 @@ export default function TextDiffPage() {
   function handleChange2(event: React.ChangeEvent<HTMLInputElement>) {
     setInput2(event.target.value);
   }
+  function handleSelectChange(event: SelectChangeEvent<any>) {
```
Comment (dlford) [inline pages/text-diff.tsx:107]: Don't use `any`, please run `npm run cq` to check for compiler/linter errors and address anything that comes up.

File: types/text-diff-types.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { Change } from 'diff';
+
+export interface DiffOptions {
```
Comment (dlford) [inline types/text-diff-types.tsx:3]: Since this is only really used in one file, lets move it to the file it is used in please

**Review by dlford — CHANGES_REQUESTED** [review]
@NirvanaPhX we're off to a good start here but there are some code quality checks failing, you can check the details at the bottom of this thread, and the Netlify deploy log in the comment at the top from Netlify bot. Most of these can be fixed or reported using the `npm run cq` script.

File: __TESTS__/text-diff.spec.tsx (hunk 4 above)
Comment (NirvanaPhX) [reply __TESTS__/text-diff.spec.tsx:56]: Tried this one but it doesn't work unfortunately.

File: pages/text-diff.tsx (hunk 5 above)
Comment (NirvanaPhX) [reply pages/text-diff.tsx:7]: For some reasons, I don't see an error here when I run `npm run lint`.

Comment (dlford) [reply pages/text-diff.tsx:7]: It might have just fixed it for you instead of showing an error

File: __TESTS__/text-diff.spec.tsx (hunk 4 above)
Comment (dlford) [reply __TESTS__/text-diff.spec.tsx:56]: Weird, lets push the other changes without writing any new tests and I'll take a look at it to see if I can figure out the issue.

Comment (NirvanaPhX) [reply __TESTS__/text-diff.spec.tsx:56]: Will do. I went through StackOverflow and it seems many people were having issues with testing MUI Select elements. I tried several ways mentioned in the posts but nothing has worked for me so far.

File: __TESTS__/text-diff.spec.tsx [hunk 8]
```diff
@@ -27,4 +27,35 @@ describe('TextDiff', () => {
...
+        )
+        .join();
+
+    userEvent.selectOptions(
+      screen.getByLabelText(/Choose Diff Options/i),
+      ['Characters Ignore Case'],
+    );
+
+    const user = userEvent.setup();
+    render(<TextDiff />);
```
Comment (dlford) [inline __TESTS__/text-diff.spec.tsx:48]: You have to render the component before you can query against it.

I generally try to keep the user event setup and render at the top of each test case. Give that a shot and it should work for you.

File: pages/text-diff.tsx [hunk 9]
```diff
@@ -2,18 +2,99 @@ import ClearIcon from '@mui/icons-material/Clear';
...
+  // {
+  //   label: 'Arrays',
+  //   description:
+  //     'Two arrays, comparing each item for strict equality (===).',
+  //   value: Diff.diffArrays,
+  // },
+];
 export default function TextDiffPage() {
   const supportsClipboardRead = useSupportsClipboardRead();
+  const [selectedOptions, setSelectedOptions] = useState<
```
Comment (dlford) [inline pages/text-diff.tsx:94]: Lets use `useLocalState` insetad of `useState` here to keep the option saved locally

**Review by dlford — CHANGES_REQUESTED** [review]
@NirvanaPhX Looking better! I think I figured out your testing issue, let me know if it works for you.

File: __TESTS__/text-diff.spec.tsx (hunk 8 above)
Comment (NirvanaPhX) [reply __TESTS__/text-diff.spec.tsx:48]: Yea that worked perfectly. Sorry, it was such a dumb mistake. Thanks a lot!

Comment (dlford) [reply __TESTS__/text-diff.spec.tsx:48]: No worries! We all miss things like that from time to time!

**Comment by NirvanaPhX** [comment]
Hi @dlford , I have included some test cases. However, when I am writing test cases for diffJsons. In the diff docs, it claims that order doesn't matter. In my case, it seems like it is affecting the result. If you wouldn't mind please take a look and let me know if I am doing anything wrong. Thanks again!

**Comment by dlford** [comment]
@NirvanaPhX No worries, I'll take a look when I get some time.

**Comment by dlford** [comment]
@NirvanaPhX Here are a few things I've found:

1. You need to [escape](https://testing-library.com/docs/ecosystem-user-event/#keyboardtext-options) the `{` character on the inputs when using `user.type`.
    ```js
    const testInputT1 = `{{"name": "Alex","salary": 50000,"ta": null}`;
    const testInputT2 = `{{"salary": 50000,"ta": null,"name": "Alex"}`;
    ```
2. I'm not sure, but maybe you don't need to `JSON.parse` the inputs before passing to `diff`? Maybe you've already tried that. In which case lets create an issue on the `diff` repo and we'll create a new issue here as well to track it, then you can just comment out that test for the time being.
3. A minor detail, but I think if we need to show the user an error it's best to do that on the page rather than in a toast, mainly because it just keeps popping up as you type. This may not matter if you don't have to use `JSON.parse` though.

Hope that helps!

**Comment by NirvanaPhX** [comment]
> @NirvanaPhX Here are a few things I've found:
> 
>     1. You need to [escape](https://testing-library.com/docs/ecosystem-user-event/#keyboardtext-options) the `{` character on the inputs when using `user.type`.
>        ```js
>        const testInputT1 = `{{"name": "Alex","salary": 50000,"ta": null}`;
>        const testInputT2 = `{{"salary": 50000,"ta": null,"name": "Alex"}`;
>        ```
> 
>     2. I'm not sure, but maybe you don't need to `JSON.parse` the inputs before passing to `diff`? Maybe you've already tried that. In which case lets create an issue on the `diff` repo and we'll create a new issue here as well to track it, then you can just comment out that test for the time being.
> 
>     3. A minor detail, but I think if we need to show the user an error it's best to do that on the page rather than in a toast, mainly because it just keeps popping up as you type. This may not matter if you don't have to use `JSON.parse` though.
> 
> 
> Hope that helps!

Thanks Dan.
1. I will try escaping the {.
2. And yes, I have tried without parsing it to a JSON object which doesn't work.
3. That sounds good. I will make the change.

**Comment by dlford** [comment]
@NirvanaPhX Cool, I will look at that JSON mode again and see what I can figure out.

**Comment by dlford** [comment]
@NirvanaPhX I think the infinite loop is happening in the `useEffect` hook, if I remove the step of clicking the JSON diff option it runs fine (and fails, as it should), but in the browser if you click the JSON diff option and then click another tool in the menu nothing happens, so I would look there and see what is causing the issue. You might need two `useEffect` hooks, because you are changing values that it is dependent on in the hook, so that might be the problem?

**Comment by NirvanaPhX** [comment]
> @NirvanaPhX I think the infinite loop is happening in the `useEffect` hook, if I remove the step of clicking the JSON diff option it runs fine (and fails, as it should), but in the browser if you click the JSON diff option and then click another tool in the menu nothing happens, so I would look there and see what is causing the issue. You might need two `useEffect` hooks, because you are changing values that it is dependent on in the hook, so that might be the problem?

I will look into it and see if I can figure it out. Again thanks for all your time, help and great advises! I really appreciate it.

**Comment by dlford** [comment]
No, thank _you_ for your time and effort!

**Comment by dlford** [comment]
@NirvanaPhX Great job!

One question; is it possible to update the diff when you change the mode selector? Not super important but I think it would be a nice detail if we could.

Everything else looks good!

**Comment by NirvanaPhX** [comment]
> @NirvanaPhX Great job!
> 
> One question; is it possible to update the diff when you change the mode selector? Not super important but I think it would be a nice detail if we could.
> 
> Everything else looks good!

Good point. Now it should be fixed. Thanks.

**Comment by dlford** [comment]
@NirvanaPhX Nice! I think you might need to update a test? Looks like one failed.

**Comment by NirvanaPhX** [comment]
> @NirvanaPhX Nice! I think you might need to update a test? Looks like one failed.

@dlford  Hi Dan, are there more tests with the ci testing during deployment? When I ran `npm run cq` locally all the tests passed and there are only 42 tests, 11 test suites but 51 tests, 13 suites when pushing to GitHub.
image
image

**Comment by dlford** [comment]
@NirvanaPhX I see what's happening, it's just taking too long and hitting the timeout, I'll fix that.

Great work on this, thank you!

**Comment by dlford** [comment]
Resolves #47

**Comment by NirvanaPhX** [comment]
> @NirvanaPhX I see what's happening, it's just taking too long and hitting the timeout, I'll fix that.
> 
> Great work on this, thank you!

Thank you so much for all the coaching and tips! 
I have learned a lot!
I will look through the issues and see if I can contribute more on this project.

**Comment by dlford** [comment]
@NirvanaPhX Awesome, feel free to create an issue if you have a new idea as well!

**Pull request merged**